src/persist/src/workload.rs

```diff
@@ -0,0 +1,268 @@
+    pub fn gen_batch(&mut self, batch_idx: usize) -> Option<ColumnarRecords> {
+        let batch_start = self.batch_size * batch_idx;
+        let batch_end = cmp::min(batch_start + self.batch_size, self.record_count);
+        if batch_start >= batch_end {
+            return None;
+        }
+        let mut batch = ColumnarRecordsBuilder::default();
+        batch.reserve(
+            batch_end - batch_start,
+            self.record_size_bytes - TS_DIFF_GOODPUT_SIZE,
+            0,
+        );
+        for record_idx in batch_start..batch_end {
+            batch.push(self.gen_record(record_idx));
+        }
+        Some(batch.finish())
+    }
+
+    fn gen_record(&mut self, record_idx: usize) -> ((&[u8], &[u8]), u64, isize) {
+        assert!(record_idx < self.record_count);
+        assert!(self.record_size_bytes > TS_DIFF_GOODPUT_SIZE);
+
+        self.key_buf.clear();
+        let key_len = self.record_size_bytes - TS_DIFF_GOODPUT_SIZE;
+        write!(&mut self.key_buf, "{:01$}", record_idx, key_len)
+            .expect("write to Vec is infallible");
+        self.key_buf.truncate(key_len);
+        assert_eq!(self.key_buf.len(), key_len);
+
+        self.val_buf.clear();
+
+        let ts = u64::cast_from(record_idx);
+        let diff = 1;
+        ((&self.key_buf, &self.val_buf), ts, diff)
+    }
+
+    /// Returns an [Iterator] of all records in batches.
+    pub fn batches(&self) -> DataGeneratorBatchIter {
+        DataGeneratorBatchIter {
+            config: self.clone(),
+            batch_idx: 0,
+        }
+    }
+
+    /// Returns an [Iterator] of all records.
+    pub fn records(&self) -> impl Iterator<Item = ((Vec<u8>, Vec<u8>), u64, isize)> {
+        let mut config = self.clone();
+        (0..self.record_count).map(move |record_idx| {
+            let ((k, v), t, d) = config.gen_record(record_idx);
+            ((k.to_vec(), v.to_vec()), t, d)
+        })
+    }
+}
+
+/// An implementation of [Iterator] for [DataGenerator].
+#[derive(Debug)]
+pub struct DataGeneratorBatchIter {
+    config: DataGenerator,
+    batch_idx: usize,
+}
+
+impl Iterator for DataGeneratorBatchIter {
+    type Item = ColumnarRecords;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let batch_idx = self.batch_idx;
+        self.batch_idx += 1;
+        self.config.gen_batch(batch_idx)
+    }
+}
+
+/// Load the dataset into a persist collection.
+///
+/// The data is loaded with one concurrent write call per generated data batch.
+/// If `do_seals` is true, data is sealed as it's loaded.
+///
+/// The "goodput" bytes represented by this data load are returned on success.
+pub fn load(
+    handle: &StreamWriteHandle<Vec<u8>, Vec<u8>>,
+    data: &DataGenerator,
+    do_seals: bool,
+) -> Result<u64, Error> {
+    let mut writes = Vec::new();
+    let mut seals = Vec::new();
+    for batch in data.batches() {
+        // It's unfortunate to turn this into a Vec just to turn it back into a
```

Comment:
it really feels like we should be able to have an abstraction that can take slices or owned vecs because iiuc we have the ability to construct columnarrecords out of either. I tried to get this to work and couldn't today. oh well. not worth holding up this pr